server/src/main/java/org/opensearch/common/compress/CompressionHelper.java

```diff
@@ -0,0 +1,99 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.common.compress;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.apache.logging.log4j.message.ParameterizedMessage;
+import org.opensearch.cluster.ClusterState;
+import org.opensearch.cluster.Diff;
+import org.opensearch.cluster.node.DiscoveryNode;
+import org.opensearch.common.bytes.BytesReference;
+import org.opensearch.common.io.stream.BytesStreamOutput;
+import org.opensearch.common.io.stream.InputStreamStreamInput;
+import org.opensearch.common.io.stream.NamedWriteableAwareStreamInput;
+import org.opensearch.common.io.stream.NamedWriteableRegistry;
+import org.opensearch.common.io.stream.OutputStreamStreamOutput;
+import org.opensearch.common.io.stream.StreamInput;
+import org.opensearch.common.io.stream.StreamOutput;
+import org.opensearch.common.io.stream.Writeable;
+import org.opensearch.transport.BytesTransportRequest;
+
+import java.io.IOException;
+
+/**
+ * A helper class to utilize the compressed stream.
+ */
+public class CompressionHelper {
+    private static final Logger logger = LogManager.getLogger(CompressionHelper.class);
+
+    /**
+     * Serialize the given cluster state or diff. It'll always use compression before writing on a newly created output
+     * stream.
+     *
+     * @param writer             Object which is going to write the content
+     * @param node               version of cluster node
+     * @param isFullClusterState flag used at receiver end to make intelligent decisions. For example, ClusterState
+     *                           assumes full state of diff of the states based on this flag.
+     * @return reference to serialized bytes
+     * @throws IOException if writing on the compressed stream is failed.
+     */
+    public static BytesReference serializeClusterState(Writeable writer, DiscoveryNode node, boolean isFullClusterState)
+        throws IOException {
+        try {
+            final BytesStreamOutput bStream = new BytesStreamOutput();
```

Comment:
Ensure you handle closing streams

Reply:
Yes, it's the same code which is already being used in PublicationTransportHandler. We can't close stream until it's written on the transport layer. Here only ByteReference is taken as output. So, later in code flow, when it's written on transport channel using the same byte reference, it'll be released.